src/key_io.cpp

```diff
@@ -43,6 +44,65 @@ class DestinationEncoder
     std::string operator()(const CNoDestination& no) const { return {}; }
 };
 
+static uint8_t GetTypecode(const void* ua, size_t index)
+{
+    return std::visit(
+        TypecodeForReceiver(),
+        reinterpret_cast<const libzcash::UnifiedAddress*>(ua)->GetReceiversAsParsed()[index]);
+}
+
+class DataLenForReceiver {
+public:
+    DataLenForReceiver() {}
+
+    size_t operator()(const libzcash::SaplingPaymentAddress &zaddr) const { return 43; }
+    size_t operator()(const libzcash::P2SHAddress &p2sh) const { return 20; }
+    size_t operator()(const libzcash::P2PKHAddress &p2pkh) const { return 20; }
+    size_t operator()(const libzcash::UnknownReceiver &unknown) const { return unknown.data.size(); }
+};
+
+static size_t GetReceiverLen(const void* ua, size_t index)
+{
+    return std::visit(
+        DataLenForReceiver(),
+        reinterpret_cast<const libzcash::UnifiedAddress*>(ua)->GetReceiversAsParsed()[index]);
+}
+
+class CopyDataForReceiver {
+    unsigned char* data;
```

Comment:
How about making this a `std::vector<unsigned char>`, so that passing in a buffer of the wrong length can be made to reliably abort rather than cause undefined behaviour?